EtherKit/Models/TypedData.swift

```diff
@@ -0,0 +1,394 @@
+    }
+
+    var subTypes: [String] = mainType.getSubTypes()
+    var seenTypes: [String] = []
+    while subTypes.count > 0 {
+      let subName = subTypes.removeFirst()
+      if seenTypes.contains(subName) {
+        continue
+      }
+      seenTypes.append(subName)
+
+      guard let subType = types[subName] else {
+        throw EtherKitError.web3Failure(reason: .parsingFailure)
+      }
+      subTypes.append(contentsOf: subType.getSubTypes())
+    }
+
+    var encodedTypes = "\(typeName)\(mainType.encodeProperties())"
+    seenTypes = seenTypes.sorted()
+    for seenName in seenTypes {
+      guard let subType = types[seenName] else {
+        throw EtherKitError.web3Failure(reason: .parsingFailure)
+      }
+      encodedTypes.append("\(seenName)\(subType.encodeProperties())")
+    }
+    return Data(bytes: Array(encodedTypes.utf8)).sha3(.keccak256)
+  }
+}
+
+public struct Domain: Unmarshaling {
+  public var name: String?
+  public var version: String?
+  public var chainId: BigUInt?
+  public var verifyingContract: Address?
+  public var salt: Data?
+
+  public init(object: MarshaledObject) throws {
+    name = try object.value(for: "name")
+    version = try object.value(for: "version")
+
+    if let uint: UInt = try? object.value(for: "chainId") {
+      chainId = BigUInt(uint)
+    } else if let string: String = try? object.value(for: "chainId"),
+      let bigUInt = BigUInt(string) {
+      chainId = bigUInt
+    }
+
+    verifyingContract = try object.value(for: "verifyingContract")
+
+    if let string: String = try? object.value(for: "salt") {
+      salt = Data(hex: string)
+    }
+  }
+
+  public func encodeData() throws -> Data {
+    var data = Data()
+    if let name = name {
+      data.append(Data(bytes: Array(name.utf8)).sha3(.keccak256))
+    }
+    if let version = version {
+      data.append(Data(bytes: Array(version.utf8)).sha3(.keccak256))
+    }
+    if let chainId = chainId {
+      let valueData = chainId.serialize()
+      let padding = 32 - valueData.count
+      data.append(Data(repeating: 0, count: padding))
+      data.append(valueData)
+    }
+    if let verifyingContract = verifyingContract {
+      let uint = BigUInt(verifyingContract.data)
+      let uintData = uint.serialize()
+      let padding = 32 - uintData.count
+      data.append(Data(repeating: 0, count: padding))
+      data.append(uintData)
+    }
+    if let salt = salt {
+      data.append(salt)
+    }
+    return data
+  }
+}
+
+public enum MessageValue {
+  case bool(value: Bool)
+  case uint(size: Int, value: BigUInt)
+  case int(size: Int, value: BigInt)
+  case address(value: Address)
+  case string(value: String)
+  case bytes(count: UnformattedDataMode, value: Data)
+  case array(count: UnformattedDataMode, value: [MessageValue])
+  case object(type: String, value: MessageObject)
+
+  public static func getMessageValue(from object: Any, type: String, types: Types) throws -> MessageValue {
```

Comment:
It might be nicer to split this up into two passes:

1. Tokenization: transforming uint[32] into (type: RawType("uint"), OpenBrace(), Number(32), CloseBrace()
2. Parsing: into .array(count: 32, value: [.uint(size: 64: value: blah)])

... but your call.

I find this section a bit difficult to read, we could even make it a bit intermediary by taking some liberties:
let type = `type.takeUntil(char => isNumber(char) || char == "[")` will give you the type you can then do a switch on.

... or just by introducing some whitespace in between sequences.  LIterally everything here is just my Tuesday morning rambling, take or leave whatever.

then rest = type.subString(after: type.count).split(separatedBy: ["[", "]") which you can parse pretty easily.
if rest.count == 3 { // is Array } if Int(rest) { // is constrained type }